picard/acousticbrainz/__init__.py

```diff
@@ -0,0 +1,162 @@
+import json
+from tempfile import NamedTemporaryFile
+
+from PyQt5 import QtCore
+
+from picard.acousticbrainz.extractor import (
+    check_extractor_version,
+    precompute_extractor_sha,
+)
+from picard.util import (
+    find_executable,
+    load_json,
+)
+from picard.util.thread import run_task
+from picard.webservice import ratecontrol
+
+
+ACOUSTICBRAINZ_HOST = "acousticbrainz.org"
+ACOUSTICBRAINZ_PORT = 80
+
+ratecontrol.set_minimum_delay((ACOUSTICBRAINZ_HOST, ACOUSTICBRAINZ_PORT), 1000)
+
+acousticbrainz_extractor = None
+acousticbrainz_extractor_sha = None
+acousticbrainz_extractor_version = None
+
+
+def ab_setup_extractor(tagger):
+    global acousticbrainz_extractor, acousticbrainz_extractor_sha, acousticbrainz_extractor_version
+    tagger.log.debug("Checking up AcousticBrainz availability")
+    acousticbrainz_extractor = find_executable("streaming_extractor_music")
+    if acousticbrainz_extractor:
+        acousticbrainz_extractor_sha = precompute_extractor_sha(acousticbrainz_extractor)
+        acousticbrainz_extractor_version = check_extractor_version(acousticbrainz_extractor, tagger)
+        if acousticbrainz_extractor_version:
+            tagger.log.debug("AcousticBrainz is available")
+            return True
+    tagger.log.warning("AcousticBrainz is not available")
+    return False
+
+
+def ab_feature_extraction(tagger, recording_id, input_path, extractor_callback):
+    # Fetch existing features from AB server to check for duplicates before extracting
+    tagger.webservice.get(
+        ACOUSTICBRAINZ_HOST,
+        ACOUSTICBRAINZ_PORT,
+        "/%s/low-level" % recording_id,
+        partial(run_extractor, input_path, extractor_callback),
+        priority=True,
+        important=False,
+        parse_response_type=None
+    )
+
+
+def run_extractor(input_path, extractor_callback, response, reply, error):
+    duplicate = False
+    # Check if AcousticBrainz server answered with the json file for the recording id
+    if not error:
+        # If it did, mark file as duplicate and skip extraction
+        try:
+            load_json(response)
+            duplicate = True
+        except json.JSONDecodeError:
+            pass
+
+    if duplicate:
+        # If an entry for the same recording ID already exists
+        # using the same extractor version exists, skip extraction
+        results = (None, 0, "Duplicate")
+        extractor_callback(results, None)
+    else:
+        # Run extractor on a different thread and call the callback when done
+        run_task(partial(extractor, input_path), extractor_callback)
+
+
+def extractor(input_path):
+    # Create a temporary file with AcousticBrainz output
+    output_file = NamedTemporaryFile("w", suffix=".json")
+    output_file.close()  # close file to ensure other processes can write to it
+
+    # Call the features extractor and wait for it to finish
+    process = QtCore.QProcess()
+    process.setProperty('picard_finished', False)

```

Comment:
We need this only if we plan to read it later. I think the reason why this is used when calling fpcalc is that it registers both finished and error signals, and looks like there are cases where both get called.

I wonder whether it would be a better alternative here instead of `run_task` only use the `QProcess` with finished and error signals connected? After all we spawn a separate process anyway, so not really a need to keep a thread waiting. Or do I miss something?

Reply:
Ended up using subprocess.run instead of QProcesses, but we could do that if I manage to discover why it is failing to run the mock extractor.